### ISSUE TITLE
[ENH] A simple Moving Average Forecaster

### ISSUE THREAD

**Issue by ngupta23** [sktime/sktime#3992]
**Is your feature request related to a problem? Please describe.**

A simple moving average forecaster is often a good baseline model. I would like this forecaster using sktime, e.g. one which will forecast based on the moving average of the last N data points. For some reason, I am not able to create this currently, so any guidance on this would be appreciated.

**Describe the solution you'd like**
See above

**Describe alternatives you've considered**

This is what I have done so far. 

1. Created a `WindowSummarizer` that find the moving average
2. Used this `WindowSummarizer` in a TransformedTargetForecaster along with `make_reduction`.
3. I was not sure what estimator to use in `make_reduction` since I do not want the output to be exactly equal to the MA value.  So I created a custom sklearn estimator which just returns X back as the prediction (in our case, X would be the MA created by the `WindowSummarizer`.

The code is below, but there are 2 issues

1. This seems to be an overly complicated implementation for a Moving Average Forecaster.
2. It does not work completely. The first prediction is correct (MA of the last 3 data points), but all subsequent predictions are not correct. It seems as if the recursive nature is not working with the way I am using the code now (or maybe a bug?)

```python

# Custom scikit estimator where predictions are the same as X (first column)

import numpy as np
from sklearn.base import BaseEstimator
from sklearn.utils.validation import check_X_y, check_array, check_is_fitted

class PredictX0(BaseEstimator):
    def fit(self, X, y):
        X, y = check_X_y(X, y, accept_sparse=True)
        self.is_fitted_ = True
        # `fit` should always return `self`
        return self

    def predict(self, X):
        X = check_array(X, accept_sparse=True)
        check_is_fitted(self, "is_fitted_")
        # Returns the first column
        return X[:, 0]

X = np.array([1, 2, 3, 4, 5]).reshape(-1, 1)
y = np.array([100, 100, 100, 100, 100])

model = PredictX0()
model.fit(X, y)
p = model.predict(X + 2)

# Predictions are the same as the X value passed in
```

```python
# Code for Moving Average forecaster

data = load_airline()
ma_window = 3
kwargs = {"lag_feature": {"mean": [[0, ma_window]]}}
transformer = WindowSummarizer(**kwargs, truncate="bfill")

forecaster = TransformedTargetForecaster(
    [
        ("ma", transformer),
        (
            "forecast",
            make_reduction(
                estimator=PredictX0(),
                scitype="tabular-regressor",
                window_length=1,
                strategy="recursive",                
            ),
        ),
    ]
)
forecaster.fit(y=data, fh=[1, 2, 3, 4, 5])
preds = forecaster.predict()
```

The last few data points and the predictions are as follows

![image](https://user-images.githubusercontent.com/33585645/209446109-c6f19076-0d29-4ea8-8471-1125a2c43149.png)

As can be seen the first prediction of 427.667 is correct, but all further predictions are simply matching the first one which is incorrect.
Kindly advice on how to implement this using sktime. Thanks!

**Additional context**

Sort of related ... I built my code based on discussions in the Issue: https://github.com/sktime/sktime/issues/1577

**Comment by fkiraly** [comment]
Re moving average:

You can get moving average as a special case of ARIMA, by switching off the AR and I parts. MA stands for moving average.

Use any of the ARIMA models in `sktime` with `p=0` and `d=0`. The `q` parameter determines the order of the moving average. More precisely, ARIMA (0, 0, q) = MA(q).

**Comment by fkiraly** [comment]
Re reducer: it seems there may be something wrong, possibly a bug. FYI @danbartl. Will also have a look.

**Comment by ngupta23** [comment]
> You can get moving average as a special case of ARIMA, by switching off the AR and I parts. MA stands for moving average.
> 
> Use any of the ARIMA models in `sktime` with `p=0` and `d=0`. The `q` parameter determines the order of the moving average. More precisely, ARIMA (0, 0, q) = MA(q).

Hi @fkiraly , This may not work for 2 reasons.

1. ARIMA MA part relies on model errors from past points and not the raw y values ([Ref](https://otexts.com/fpp3/MA.html)).
![image](https://user-images.githubusercontent.com/33585645/209449355-b16be36c-5d49-4ead-93da-affcd067ea57.png)

Hence, after crossing future points > q, it will not be able to predict ([Link to Code](https://gist.github.com/ngupta23/4aaab9205ef3de3921c3a5a580f7701d)). 

![image](https://user-images.githubusercontent.com/33585645/209449181-f753c180-de06-4fd6-b277-1a6a9e886c4e.png)

2. Even if it was working,, it does not provide any control over the coefficients of the model so the past 'q' points would be weighted differently (theta values in the equation above) depending on how the model gets fitted.

I am looking for a Simple Moving Average where all past q points are weighted equally. This would be a starting (baseline) model, then I would add additional complexity on top of that iteratively to see what helps to improve the model. This could include for example, building more features beyond a simple moving average (e.g. lags, standard deviations, mins, max, etc) or changing out the regressor to more complex ones so see if they help. Hence, a working baseline solution using regressors would be helpful to have. 

Thanks!

**Comment by fkiraly** [comment]
oh, I see - you want an *equal weighted moving average* that is recursively applied?

**Comment by ngupta23** [comment]
> oh, I see - you want an _equal weighted moving average_ that is recursively applied?

Yes, a simple Moving Average, similar to `pd.rolling(ma_window).mean()`, but applied recursively the forecasts as far out into the future as needed.

**Comment by fkiraly** [comment]
>but applied recursively to the forecasts as far out into the future as needed.

Well, it will converge to a constant if you go too far in the future...

**Comment by ngupta23** [comment]
> Well, it will converge to a constant if you go too far in the future...

It may, but I think only if all the values in a future Moving Average window are the same value. It would be based on pure luck or due to the choice of a very small MA window.

**Comment by fkiraly** [comment]
> It may, but I think only if all the values in a future Moving Average window are the same value. It would be based on pure luck or due to the choice of a very small MA window.

No, I think: it always will, for all lengths of MA windows, and all values of the series, eventually.

Proof: let $m$ be the window length, and $a_1, \dots, a_m$ be the last $m$ values of the series. Let $a_{m+1}, \dots$ be the future values obtained recursively through the recurrence relation $a_j := \frac{1}{m}\sum_{i=1}^q a_{j-i}$.

The recurrence is linear, so you have $(a_{1+n}, \dots, a_{m+n}) = A^n (a_1, \dpts, a_m)$ for the $(m \ times m)$ square matrix which has $\frac{1}{m}$ in the last row, otherwise zeros in the first column, and the rest is an identity matrix (ones on the off-diagonal and zeros otherwise).

This is a special Toeplitz matrix or a special convolution matrix, which can be verified to have one eigenvalue 1, and the remaining eigenvalues are possibly complex but norm strictly smaller than one.

The asmyptotic value ends up being interesting:
$\frac{\sum_{i=1}^m i\cdot a_i}{\sum_{i=1}^m i}$,
i.e., the weighted mean of the original $a_i$ where the weight of $a_i$ is the integer $i$.

**Comment by ngupta23** [comment]
> No, I think: it always will, for all lengths of MA windows, and all values of the series, eventually.

The key there is "eventually". For long horizons, that is correct (similar to what ARMA models would produce). But for short horizons, they can still produce reasonable baseline results. The key there is the selection of an appropriate window size (not too short as that can result in constant forecasts even in the short term). The definition of "short term" of course is business dependent and will vary based on frequency of data (monthly vs. weekly) and so will the appropriate selection of the MA window size.

![image](https://user-images.githubusercontent.com/33585645/209472450-3d10c80b-4d58-47f8-bfce-03dd5aa3c850.png)

But I think we are digressing from the original intent of this request. I would like to see if this capability can be implemented with the reduced regression setting in sktime today. It would be very helpful for my use case.

Thanks!

**Comment by fkiraly** [comment]
I think what you need to do with the present framework is to use a different tabular predictor, namely the hard coded linear model that averages:

```python
class AverageX(BaseEstimator):
    def fit(self, X, y):
        X, y = check_X_y(X, y, accept_sparse=True)
        self.is_fitted_ = True
        # `fit` should always return `self`
        return self

    def predict(self, X):
        X = check_array(X, accept_sparse=True)
        check_is_fitted(self, "is_fitted_")
        # Returns the first column
        return X.mean(axis=1)
```

Then, using recursive reduction:
```python
make_reduction(
    estimator=AverageX(),
    scitype="tabular-regressor",
    window_length=whatever_you_want,
    strategy="recursive",                
),
```

**Comment by ngupta23** [comment]
Thanks @fkiraly. I will try it out and let you know how it goes.

**Comment by ericjb** [comment]
A simple moving average forecaster is available in sktime. It is a special case of the NaiveForecaster.  e.g.

NaiveForecaster(strategy='mean', sp=1, window_length=12)

Here is a fleshed out example to show how you would use it:

```
import numpy as np
from sktime.datasets import load_airline
from sktime.forecasting.naive import NaiveForecaster
from sktime.utils import plot_series
import matplotlib.pyplot as plt

y = load_airline()
fc = NaiveForecaster(strategy='mean', sp=1, window_length=12)
fc.fit(y)
fitted = fc.predict(y.index)
pred = fc.predict(np.arange(1,13))

_, _ = plot_series(y, fitted, pred)
plt.show()
```

This [ENH] request can be closed out.

**Comment by fkiraly** [comment]
@ericjb, you are right - we could add an alias for the naive forecaster, or just close this issue, any preference?

**Comment by ericjb** [comment]
I think an alias is better so that users who are looking for such a forecaster can find it (assuming that adding an alias is not a lot of work.)

**Comment by fkiraly** [comment]
ok, I leave the issue open with the aim to produce an alias.

**Comment by marrov** [comment]
> A simple moving average forecaster is available in sktime. It is a special case of the NaiveForecaster.

@ericjb - This is actually not true. It's a bit of semantics, but what OP was mentioning is the recursive moving average (sometimes called simple moving average in the demand forecasting literature) that @fkiraly produced a quick code snippet for.  A comparison can be seen below:

```python
from sklearn.base import BaseEstimator
from sklearn.utils.validation import check_array, check_is_fitted, check_X_y
from sktime.datasets import load_airline
from sktime.forecasting.base import BaseForecaster
from sktime.forecasting.compose import make_reduction
from sktime.forecasting.naive import NaiveForecaster
from sktime.split import temporal_train_test_split
from sktime.utils import plot_series

max_fh = 12
fh = range(1, max_fh + 1)
y = load_airline()
y_train, y_test = temporal_train_test_split(y, test_size=max_fh)
ma = NaiveForecaster(strategy="mean", sp=1, window_length=max_fh)
ma.fit(y_train)
y_pred_ma = ma.predict(fh)


def _create_recursive_moving_average_model(window: int) -> BaseForecaster:
    """
    Create a recursive moving average (rma) model with the given window.

    Surprisingly, this simple model is not available in sktime by default. Note that
    this is different to a moving average model for a fh > 1 as it will use the forecast
    of fh=1 for fh=2 and so on.

    See more at: https://github.com/sktime/sktime/issues/3992
    """

    class AverageX(BaseEstimator):
        def fit(self, X, y):
            X, y = check_X_y(X, y, accept_sparse=True)
            self.is_fitted_ = True
            return self

        def predict(self, X):
            X = check_array(X, accept_sparse=True)
            check_is_fitted(self, "is_fitted_")
            return X.mean(axis=1)

    recursive_moving_avg_model = make_reduction(
        estimator=AverageX(),
        scitype="tabular-regressor",
        window_length=window,
        strategy="recursive",
    )
    return recursive_moving_avg_model


rma = _create_recursive_moving_average_model(window=max_fh)
rma.fit(y_train)
y_pred_rma = rma.predict(fh)

_, _ = plot_series(
    y, y_pred_ma, y_pred_rma, labels=["y", "naive mean", "recursive moving average"]
)
```

The plot should look like this.

<img width="1315" height="351" alt="Image" src="https://github.com/user-attachments/assets/6f95fde3-7c04-4e57-bf60-e2f071986980" />

Actually, this is quite relevant for a simple baseline in the demand forecasting literature, where it is common to see authors claiming that beating a SMA3 model (recursive moving average of the last 3 periods) is surprisingly difficult to do. I can provide some refs if needed.

I have actually been using this snippet since forever so I think having a forecaster for this would be good. I'll mark it as a good first issue for a new contributor.

**Comment by ericjb** [comment]
@marrov you are correct that what I suggested was non-recursive and not what the OP requested. (Although it is what is often meant by the term "moving average forecasts".)

The suggested implementations by you and @fkiraly using recursive tabular-regressor would address the request.

A different approach would be to observe that a recursive moving average is a special case of an AR process where the AR coefficients are all 1/N, where N is the number of elements averaged. (phi_1 = phi_2 = ... = phi_N = 1/N).

We can consider an AR process as a special case of an ARIMA(p,d,q), where d=q=0. 
(N.B. @fkiraly had this backwards in his suggestion from Dec. 24, 2022, where he set p=d=0.)

Along these lines, as a "first project" I would suggest the more general ability to generate simulated paths of an AR process (or more generally an ARIMA process). This seems to be a "hole" in sktime (and, surprisingly, also in statsmodels). 

In R, it is possible to generate ARIMA paths. The functionality might fall out of my PR #8538 (perhaps with a bit of extra work). Unfortunately, that PR seems to be a bit stuck.

**Comment by archittmittal** [comment]
Hi @marrov  @fkiraly  👋

I'd like to work on this issue to implement the RecursiveMovingAverageForecaster.

My approach:

Create a new forecaster using [_DelegatedForecaster](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) that wraps [make_reduction](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) with strategy="recursive" and an AverageX estimator
New module at [sktime/forecasting/moving_average.py](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html)
Proper docstrings, tags, and tests
I've already started implementation. Could I be assigned?

Thanks!

**Comment by ericjb** [comment]
A few comments (essentially repeating some of my previous comment from 2 days ago.)
I think this forecaster would not be of particular interest. It was a random request from 3 years ago.
By contrast, the ability to generate simulated paths for an AR process or, more generally, an ARIMA process, could be of general interest and used in a variety of applications. 

If you build the AR (or ARIMA) simulations generator, then you can create this estimator as a one-liner (more or less) on top of the simulator. This follows from the observation that the requested estimator generates a simulated path for an AR process that has equal weights and where the initial state is the last N observations from the series y used to fit the model, where N is the window size (and where the variance of the periodic shocks is set to zero). Done.
